framework/quota/src/org/apache/cloudstack/quota/dao/QuotaBalanceDaoImpl.java

```diff
@@ -0,0 +1,201 @@
+        SearchCriteria<QuotaBalanceVO> sc = createSearchCriteria();
+        sc.addAnd("accountId", SearchCriteria.Op.EQ, accountId);
+        sc.addAnd("domainId", SearchCriteria.Op.EQ, domainId);
+        sc.addAnd("creditsId", SearchCriteria.Op.EQ, 0);
+        sc.addAnd("updatedOn", SearchCriteria.Op.LT, beforeThis);
+        List<QuotaBalanceVO> quotab = this.search(sc, filter);
+        TransactionLegacy.open(opendb).close();
+        return quotab.size() > 0 ? quotab.get(0) : null;
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public QuotaBalanceVO findLaterBalanceEntry(final long accountId, final long domainId, final Date afterThis) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        TransactionLegacy.open(TransactionLegacy.USAGE_DB).close();
+        Filter filter = new Filter(QuotaBalanceVO.class, "updatedOn", true, 0L, 1L);
+        SearchCriteria<QuotaBalanceVO> sc = createSearchCriteria();
+        sc.addAnd("accountId", SearchCriteria.Op.EQ, accountId);
+        sc.addAnd("domainId", SearchCriteria.Op.EQ, domainId);
+        sc.addAnd("creditsId", SearchCriteria.Op.EQ, 0);
+        sc.addAnd("updatedOn", SearchCriteria.Op.GT, afterThis);
+        List<QuotaBalanceVO> quotab = this.search(sc, filter);
+        TransactionLegacy.open(opendb).close();
+        return quotab.size() > 0 ? quotab.get(0) : null;
+    }
+
+    @Override
+    public void saveQuotaBalance(final List<QuotaBalanceVO> credits) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.USAGE_DB);
+        try {
+            for (QuotaBalanceVO credit : credits) {
+                persist(credit);
+            }
+        } finally {
+            txn.close();
+        }
+        TransactionLegacy.open(opendb).close();
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public List<QuotaBalanceVO> findCreditBalance(final long accountId, final long domainId, final Date lastbalancedate, final Date beforeThis) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        TransactionLegacy.open(TransactionLegacy.USAGE_DB).close();
+        Filter filter = new Filter(QuotaBalanceVO.class, "updatedOn", true, 0L, Long.MAX_VALUE);
+        SearchCriteria<QuotaBalanceVO> sc = createSearchCriteria();
+        sc.addAnd("accountId", SearchCriteria.Op.EQ, accountId);
+        sc.addAnd("domainId", SearchCriteria.Op.EQ, domainId);
+        sc.addAnd("creditsId", SearchCriteria.Op.GT, 0);
+        if ((lastbalancedate != null) && (beforeThis != null) && lastbalancedate.before(beforeThis)) {
+            sc.addAnd("updatedOn", SearchCriteria.Op.BETWEEN, lastbalancedate, beforeThis);
+        } else {
+            return new ArrayList<QuotaBalanceVO>();
+        }
+        List<QuotaBalanceVO> qb = search(sc, filter);
+        TransactionLegacy.open(opendb).close();
```

Comment:
See previous comments regarding apparently needless open/close of a database transaction.